source/extensions/filters/http/rate_limit_quota/filter.cc

```diff
@@ -201,102 +206,189 @@ RateLimitQuotaFilter::sendImmediateReport(const size_t bucket_id,
+    }
+    // Request is throttled.
+    ENVOY_LOG(trace,
+              "Request with hashed bucket_id {} is throttled by token "
+              "bucket limiter",
+              bucket_id);
+    ENVOY_LOG(debug,
+              "Denying request as token bucket is exhausted for bucket_id {}. "
+              "Initial assignment: {}.",
+              bucket_id, rate_limit_strategy.token_bucket().ShortDebugString());
+    return Envoy::Http::FilterHeadersStatus::StopIteration;
+  }
+
+  ENVOY_LOG(error,
+            "Failing open as selected bucket action for bucket_id {} contains "
+            "an unsupported rate limit strategy: {}",
+            bucket_id, rate_limit_strategy.DebugString());
+  return Envoy::Http::FilterHeadersStatus::Continue;
+}
+
+bool isCachedActionExpired(TimeSource& time_source, const Bucket& bucket) {
   // First, check if assignment has expired nor not.
   auto now = std::chrono::duration_cast<std::chrono::nanoseconds>(
-      time_source_.monotonicTime().time_since_epoch());
+      time_source.monotonicTime().time_since_epoch());
   auto assignment_time_elapsed = Protobuf::util::TimeUtil::NanosecondsToDuration(
       (now - std::chrono::duration_cast<std::chrono::nanoseconds>(
-                 quota_buckets_[bucket_id]->first_assignment_time.time_since_epoch()))
+                 bucket.current_assignment_time.time_since_epoch()))
           .count());
-  if (assignment_time_elapsed > quota_buckets_[bucket_id]
-                                    ->bucket_action.quota_assignment_action()
-                                    .assignment_time_to_live()) {
-    // If expired, remove the cache entry.
-    quota_buckets_.erase(bucket_id);
-
-    // Default strategy is fail-Open (i.e., allow_all).
-    auto ret_status = Envoy::Http::FilterHeadersStatus::Continue;
-    // Check the expired assignment behavior if configured.
-    // Note, only fail-open and fail-close are supported, more advanced expired assignment can be
-    // supported as needed.
-    if (match_action.bucketSettings().has_expired_assignment_behavior()) {
-      if (match_action.bucketSettings()
-              .expired_assignment_behavior()
-              .fallback_rate_limit()
-              .blanket_rule() == envoy::type::v3::RateLimitStrategy::DENY_ALL) {
-        sendDenyResponse();
-        ret_status = Envoy::Http::FilterHeadersStatus::StopIteration;
-      }
-    }
 
-    return ret_status;
+  return (assignment_time_elapsed >
+          bucket.cached_action->quota_assignment_action().assignment_time_to_live());
+}
+
+Http::FilterHeadersStatus
+RateLimitQuotaFilter::processCachedBucket(size_t bucket_id,
+                                          const RateLimitOnMatchAction& match_action) {
+  auto* cached_bucket = quota_buckets_[bucket_id].get();
+  Http::FilterHeadersStatus ret_status;
+
+  if (!cached_bucket->cached_action.has_value()) {
+    // If no cached action, use the default action.
```

Comment:
Is this true? if cached action expires, should we just fallback to expired assignment behavior?

Reply:
On second thought, I think your logic is as below, which make sense 
(a) first time If expired ---> 1. processExpiredBucket 2. reset `cached_action`
(b) then second time --->  2. use the default action

Reply:
If there's no cached action at all, it means that expiration has already happened in a previous run or no assignment has been received yet from the RLQS server. So yes, at this point, we're relying on the configured default.